source/server/drain_manager_impl.cc

```diff
@@ -70,28 +70,44 @@ bool DrainManagerImpl::drainClose() const {
 
   const auto remaining_time =
       std::chrono::duration_cast<std::chrono::seconds>(drain_deadline_ - current_time);
+  const auto drain_time = server_.options().drainTime();
   ASSERT(server_.options().drainTime() >= remaining_time);
-  const auto elapsed_time = server_.options().drainTime() - remaining_time;
+  const auto drain_time_count = drain_time.count();
+
+  // If the user hasn't specified a drain timeout it will be zero, so we'll go
+  // and and close immediately. Otherwise we'll use the drain timeout as a modulus
+  // to a random number to salt the drain timing.
+  if (drain_time_count == 0) {
+    return true;
+  }
+  const auto elapsed_time = drain_time - remaining_time;
   return static_cast<uint64_t>(elapsed_time.count()) >
-         (server_.api().randomGenerator().random() % server_.options().drainTime().count());
+         (server_.api().randomGenerator().random() % drain_time_count);
 }
 
 Common::CallbackHandlePtr DrainManagerImpl::addOnDrainCloseCb(DrainCloseCb cb) const {
+  ASSERT_IS_MAIN_OR_TEST_THREAD();
   ASSERT(dispatcher_.isThreadSafe());
 
   if (draining_) {
     const MonotonicTime current_time = dispatcher_.timeSource().monotonicTime();
 
     // Calculate the delay. If using an immediate drain-strategy or past our deadline, use
     // a zero millisecond delay. Otherwise, pick a random value within the remaining time-span.
-    std::chrono::milliseconds drain_delay =
-        (server_.options().drainStrategy() != Server::DrainStrategy::Immediate &&
-         current_time < drain_deadline_)
-            ? std::chrono::milliseconds(server_.api().randomGenerator().random() %
-                                        std::chrono::duration_cast<std::chrono::milliseconds>(
-                                            drain_deadline_ - current_time)
-                                            .count())
-            : std::chrono::milliseconds{0};
+    std::chrono::milliseconds drain_delay{0};
+    if (server_.options().drainStrategy() != Server::DrainStrategy::Immediate) {
+      if (current_time < drain_deadline_) {
+        auto ms =
+            std::chrono::duration_cast<std::chrono::milliseconds>(drain_deadline_ - current_time)
+                .count();
+        // Note; current_time may be less than drain_deadline_ by a microsecond, and
+        // when we round to milliseconds that might be 0, which would throw an arithmetic
+        // exception if used as a modulus.
```

Comment:
Sorry how does this happen? I'm confused how the time could ever be less?

Reply:
I'm going to add a test case to get to this spot with 1us variation on DrainManagerImplTest, RegisterCallbackAfterDrainBeginGradualStrategy.